### ISSUE TITLE
`--check` silently ignored when formatting from stdin

### ISSUE THREAD

**Issue by pyrrho** [rust-lang/rustfmt#3871]
When I run,
```sh
echo '
fn   foo (   ) {
    fake_fn("this is fine");
    fake_fn   ( "This not so much"    )    ;
}
' | rustfmt --check
```
the output is equivalent to `rustfmt --emit stdout`. Ideally, I would like to see the above call emit something like,
```diff
Diff in 'stdin' at line 1:
-fn   foo (   ) {
+fn foo() {
     fake_fn("this is fine");
-    fake_fn   ( "This not so much"    )    ;
+    fake_fn("This not so much");
 }
```
If that's not possible, a warning or error message noting the ignored option would make this behavior less surprising, and may help nudge users towards valid usage.

It might also be worth noting that `rustfmt +nightly --emit json` is silently ignored in the same situations.

```
$ rustfmt --version
rustfmt 1.4.4-stable (0462008d 2019-08-06)
```
```
$ rustfmt +nightly --version
rustfmt 1.4.9-nightly (33e36670 2019-10-07)
```

**Comment by calebcartwright** [comment]
When using stdin input, rustfmt will always use stdout emit mode (by design). However, as you noted, because of this rustfmt is silently ignoring the `--check` and `--emit` options. 

I don't think it'll be viable to support the various file emit modes with stdin (include the diff emit mode that comes with `--check` when formatting files), but would it be helpful for this to be "less silent"? 

I suppose we could have rustfmt err if you include `--check` or `--emit` with stdin (just like it does if you try to use both `--check` and `--emit`) or print a warning message, and/or we could note the stdin = stdout only in the readme

**Comment by topecongiro** [comment]
> When using stdin input, rustfmt will always use stdout emit mode (by design).

Well, we can support `--check` on stdin input it is feasible. If not, then yes, rustfmt should exit with an error.

**Comment by pyrrho** [comment]
I agree, if it's possible to emit diffs (with `--check`) or JSON (with `+nightly --emit json`), that would be the best solution.

If not, an error (as occurs with `--check --emit stdout`) would be better than silently ignoring options.

**Comment by jugglerchris** [comment]
I'd be interested in working on this - making it an error as a first step, and then looking into making `--check` and `--emit` work.

**Comment by calebcartwright** [comment]
Currently, the checkstyle and json emitters would not handle `stdin` as they are expecting a real file path:

https://github.com/rust-lang/rustfmt/blob/69c7dbcd5026b75dee78e4fd14fbe57ffd1d4e4e/src/emitter/json.rs#L40-L60

https://github.com/rust-lang/rustfmt/blob/master/src/emitter/checkstyle.rs#L23-L37

I included the file path check when I implemented the json emitter because it was there in the checkstyle emitter, though in hindsight I'm not sure if that's _really_ required

I don't see any reason why `name` couldn't  be `stdin` in the json output
```json
[  
  {  
    "name":"tests/writemode/source/json.rs",
    "mismatches":[  
      {  
        "original_begin_line":5,
        "original_end_line":7,
        "expected_begin_line":5,
        "expected_end_line":5,
        "original":"fn foo_expr() {\n    1\n}",
        "expected":"fn foo_expr() { 1 }"
      }
]
```

The diff emitter (which is used with `--check`), would need to be tweaked as well to not use  `ensure_real_path` if the file name is `stdin`, which it does when either the user includes the `-l` option (which lists file names) or if there is a newline style diff.

**Comment by jugglerchris** [comment]
I've taken the liberty of speculatively doing the first step (returning errors) in #3875; I'll probably speculatively go ahead and try to make them work too.  :-)

**Comment by jugglerchris** [comment]
I got some time to look at making the other emit modes work on [this branch](https://github.com/jugglerchris/rustfmt/tree/stdin_emit).  I haven't done any tests so far for the changes though.  Looking through, `tests/rustfmt/main.rs` looks like the most likely place to add tests of the rustfmt executable - is that right?

**Comment by calebcartwright** [comment]
> Looking through, tests/rustfmt/main.rs looks like the most likely place to add tests of the rustfmt executable - is that right?

I'd take a look at `src/test/mod.rs`, for example
https://github.com/rust-lang/rustfmt/blob/69c7dbcd5026b75dee78e4fd14fbe57ffd1d4e4e/src/test/mod.rs#L409-L429

and the in-file unit tests for the relevant emitters `src/emitter/..`
https://github.com/rust-lang/rustfmt/tree/master/src/emitter

**Comment by jugglerchris** [comment]
Thanks @calebcartwright , that was very useful.  I've raised a new PR with the JSON and checkstyle emitters now working with stdin (not yet `--check`).  I'm not happy about adding a `#[rustfmt::skip]` in the tests - is the right fix to put the outputs in text files and load them as with some other tests?

**Comment by calebcartwright** [comment]
Thanks @jugglerchris! Will respond to your question over in #3894

**Comment by jugglerchris** [comment]
I think #3894 and #3896 (which includes the changes from the other - perhaps it wasn't worth breaking them into two PRs... I can close one if that's easier) finish resolving this ticket.

**Comment by jugglerchris** [comment]
Have I missed any parts to this or do we think this can be closed?

**Comment by calebcartwright** [comment]
Was the diff emitter updated? That's the emitter used with `--check` (the checkstyle emitter is for the xml report)

**Comment by jugglerchris** [comment]
You can't select it explicitly with `--emit`, but `--check` with input from stdin works for me:
```
$ cargo +nightly run --bin rustfmt -- --check < ./foo.rs
Diff in stdin at line 1:
-fn main
-()
-
-{
-}
+fn main() {}
+

```

**Comment by calebcartwright** [comment]
> You can't select it explicitly with `--emit`, but `--check` with input from stdin works for me:
> 
> ```
> $ cargo +nightly run --bin rustfmt -- --check < ./foo.rs
> Diff in stdin at line 1:
> -fn main
> -()
> -
> -{
> -}
> +fn main() {}
> +
> ```

I know, but if the only formatting issue is line endings and/or the user also includes the `-l` option there would still be an issue when using `--check` as the assocaited emitter is still expecting the file name to be an actual path, no?

**Comment by jugglerchris** [comment]
Ok, I do get a panic with `--check -l`, and the "Incorrect newline style" message just doesn't appear with stdin.  Thanks!  I'll take a look at those.

**Comment by jugglerchris** [comment]
I think the missing "Incorrect newline style" is because the `SourceMap` doesn't remember the original line endings, so by the time it gets to `DiffEmitter::emit_formatted_file()` it's identical to the fixed output.  But I've hopefully fixed the panics now.

**Comment by calebcartwright** [comment]
👍 I believe your updates to the diff emitter will cover both cases.

It's true that internally the parser normalizes line endings to unix style, so the `original_text` value will always have unix style endings when using stdin. However, I do think there were some scenarios where `formatted_text` could have windows style endings (namely with `newline_style = "Windows"` or `newline_style = "Native"` when running on windows) which may have resulted in panics with the previous behavior.

**Comment by topecongiro** [comment]
Closed via #3910.
